CppImport/src/macro/MetaDefinitionManager.cpp

```diff
@@ -0,0 +1,249 @@
+
+		if (!result)
+		{
+			result = new OOModel::Module("notenvision");
+			root_->modules()->append(result);
+		}
+	}
+
+	Q_ASSERT(result);
+	return result;
+}
+
+OOModel::MetaDefinition* MetaDefinitionManager::metaDefinition(const clang::MacroDirective* md)
+{
+	QString h = definitionManager_->hash(md);
+
+	if (!metaDefinitions_.contains(h))
+		return nullptr;
+
+	return metaDefinitions_.value(h);
+}
+
+void MetaDefinitionManager::createMetaDef(QVector<Model::Node*> nodes, MacroExpansion* expansion, NodeMapping* mapping,
+														QVector<MacroArgumentInfo>& arguments)
+{
+	if (!metaDefinition(expansion->definition))
+	{
+		auto metaDef = new OOModel::MetaDefinition(definitionManager_->definitionName(expansion->definition));
+		metaDefinitions_.insert(definitionManager_->hash(expansion->definition), metaDef);
+
+		// add formal arguments based on the expansion definition
+		for (auto argName : clang_->argumentNames(expansion->definition))
+			metaDef->arguments()->append(new OOModel::FormalMetaArgument(argName));
+
+		auto metaDefParent = metaDefinitionParent(expansion->definition);
+
+		// check whether this expansion is not a potential partial begin macro specialization
+		if (auto beginChild = xMacroManager_->partialBeginChild(expansion))
+			xMacroManager_->handlePartialBeginSpecialization(metaDefParent, metaDef, expansion, beginChild);
+		else
+		{
+			// case: meta definition is not a partial begin macro specialization
+
+			if (nodes.size() > 0)
+			{
+				// create a new context with type equal to the first node's context
+				auto actualContext = StaticStuff::actualContext(mapping->original(nodes.first()));
+				metaDef->setContext(StaticStuff::createContext(actualContext));
+
+				// clone and add nodes to the metaDef
+				for (auto n : nodes)
+				{
+					NodeMapping childMapping;
+					auto cloned = StaticStuff::cloneWithMapping(mapping->original(n), &childMapping);
+
+					lexicalHelper_->applyLexicalTransformations(cloned, &childMapping,
+																					 clang_->argumentNames(expansion->definition));
+
+					insertChildMetaCalls(expansion, &childMapping);
+
+					if (removeUnownedNodes(cloned, expansion, &childMapping))
+						continue;
+
+					insertArgumentSplices(mapping, &childMapping, arguments);
+
+					StaticStuff::addNodeToDeclaration(cloned, metaDef->context());
+				}
+			}
+
+			// add all child expansion meta calls that are not yet added anywhere else as declaration meta calls
+			for (auto childExpansion : expansion->children)
+				if (!childExpansion->metaCall->parent())
+					metaDef->context()->metaCalls()->append(childExpansion->metaCall);
+		}
+
+		metaDefParent->subDeclarations()->append(metaDef);
+	}
+
+	// qualify the meta call
+	auto callee = DCast<OOModel::ReferenceExpression>(expansion->metaCall->callee());
+	callee->setPrefix(definitionManager_->expansionQualifier(expansion->definition));
+}
+
+void MetaDefinitionManager::renameMetaCalls(Model::Node* node, QString current, QString replace)
```

Comment:
It was required for an older solution. The current system does not even call this function any more. I removed it.